learn/regex.md

```diff
@@ -0,0 +1,348 @@
+
+The `re` module in Python provides a method called `sub()` which stands for substitute. It's the cornerstone for performing **python replaceall regex** operations. The syntax is straightforward: `re.sub(pattern, replacement, string)`. This method searches for the pattern in the given string and replaces it with the replacement string.
+
+### Example: Simple Word Replacement
+
+```python
+import re
+
+text = "Hello World! Welcome to the World of Python."
+new_text = re.sub(r"World", "Universe", text)
+print(new_text)
+```
+Output: `Hello Universe! Welcome to the Universe of Python.`
+
+This example demonstrates how to replace all instances of `World` with `Universe` in the given string.
+
+### Example: Removing Punctuation
+
+```python
+import re
+
+text = "Hello, World! How are you today?"
+clean_text = re.sub(r"[,!.?]", "", text)
+print(clean_text)
+```
+
+Output: `Hello World How are you today`
+
+Here, we efficiently remove all common punctuation marks from our string, showcasing the utility of `re.sub()` for text cleaning purposes.
+
+Mastering **Python string replace regex** with `re.sub()` function offers a flexible and powerful way to manipulate strings in Python. Its ability to search and replace patterns makes it an invaluable tool for anyone looking to perform sophisticated text processing tasks effortlessly.
+
+Remember, the key to proficiency with **Python replaceall regex** is practice. Implement these examples in your projects, and you'll soon discover even more ways to streamline your code with Python's regex capabilities.  
+  
+## Regex tester  
+
+Looking for a Python regex tester to validate and refine your regular expressions efficiently? Testing Python regex patterns is essential in data parsing, cleansing, and preparation tasks. With a Python regex tester, you can swiftly test, debug, and optimize your regular expressions to match or extract the exact data set you need.
+
+### Example: Search Emails by Pattern
+
+```python
+import re
+
+text = "Contact us at support@example.com or sales@example.net"
+pattern = r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"
+
+# Test Python regex
+matches = re.findall(pattern, text)
+
+for match in matches:
+    print("Email found:", match)
+```
+
+### Example: Search Phone Numbers by Pattern
+
+```python
+import re
+
+text = "Reach us at +1-555-123-4567 or +1-555-765-4321"
+pattern = r"\+\d{1,3}-\d{3}-\d{3}-\d{4}"
+
+# Python regex tester
+matches = re.findall(pattern, text)
+
+for match in matches:
+    print("Phone number found:", match)
+```
+
+These examples highlight the utility of a Python regex tester in extracting and verifying data patterns swiftly. Whether you're working on data scraping, validation, or preprocessing tasks, regular expressions are invaluable tools in your Python toolkit. By leveraging a Python regex tester, you can ensure your expressions are accurate, efficient, and ready for any data challenge.
+ 
+## Regex cheat sheet  
```

Comment:
Let's do not use the `Regex cheat sheet` keyword. When a user searches for a `cheat sheet` - he expects some picture or PDF file with Regex rules. We don't provide such feature. Let's rewrite this heading (+ other keyword usage if it exists in the article).